### PR TITLE
Added **kwargs to dictionary of Subject and Image

### PR DESCRIPTION
**Pull request by nwschurink**
Edited the `__init__` of `Image` and `Subjec`t such that `**kwargs` get assigned to the dictionary of the class.

I added name again as a positional argument, as the name attribute is used in several subroutines and cannot be undefined.

Reference: #112.

### CODE REVIEW

File: torchio/data/images.py [hunk 1]
```diff
@@ -87,10 +87,11 @@ class Subject(list):
         **kwargs: Items that will be added to the subject sample.
     """
 
-    def __init__(self, *images: Image, **kwargs):
+    def __init__(self, *images: Image, name: str = '', **kwargs):
```
Comment (fepegar) [inline torchio/data/images.py:90]: Can you explain why you're keeping `name`?

File: torchio/data/images.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def __init__(self, name: str, path: TypePath, type_: str, **kwargs):
         self.name = name
         self.path = self._parse_path(path)
         self.type = type_
-        self.kwargs = kwargs
```
Comment (fepegar) [inline torchio/data/images.py:44]: This attribute is referenced in `ImagesDataset` so I guess it should stay around.

Comment (fepegar) [reply torchio/data/images.py:44]: Also, if there `kwargs` has many items, users would need many lines of code to add the parameters to the sample. This way, it's just a single line.

File: torchio/data/images.py (hunk 1 above)
Comment (fepegar) [reply torchio/data/images.py:90]: I guess you mean it's used in `Subject.__repr__`. That's not very important, so we can just get rid of it.

File: torchio/data/images.py (hunk 2 above)
Comment (nwschurink) [reply torchio/data/images.py:44]: I missed the reference in ImageDataset, but what do you mean with that your need many likes of code to add parameters to a sample? Do you mean, like when you'd like a list of labels from an ImageDataset instance?

Comment (nwschurink) [reply torchio/data/images.py:44]: And what if Subject inherits from 'dict' instead of 'list'? Then you could easily iterate though Subject.keys()

Comment (fepegar) [reply torchio/data/images.py:44]: I'm not sure that would be necessary. Why not just keeping `kwargs`?

Comment (nwschurink) [reply torchio/data/images.py:44]: Hmm yea you are right it's not really necessary, maybe it's just a matter of taste then. Accessing Subject.kwargs['key'] feels a bit awkward as I would say the keyword arguments are attributes of the Subject. However, it works perfectly fine this way so it is a good sollution ;) I'll just work with it like this

Comment (fepegar) [reply torchio/data/images.py:44]: Maybe there's a misunderstanding. I propose keeping the kwargs attribute **and** doing the `Subject.__dict__.update(kwargs)`. So you can still do `subject.diagnose` or, if needed, `subject.kwargs['diagnose']`.

If `Subject` inherits from `dict`, I guess then there's no need for `Image` to have a `name`. Maybe that makes more sense. But what would be the arguments of `Subject`? A dictionary similar to the `sample_dict` would make sense, although it might be a bit more verbose than the current approach.

**Comment by fepegar** [comment]
```python
subject = Subject(
    t1=Image(path, type),
    t2=Image(path, type),
    age=25,
)

sample = {}
for key, value in subject.items():
    if isinstance(value, Image):
        sample[key] = get_image_dict_from_image(value)
    else:
        sample[key] = value
```

Maybe a similar approach could be useful for `Image` as well. But I'm afraid it will be hard to collate the batch by the `DataLoader` if keys start to differ between samples.

@romainVala, what do you think?

**Comment by romainVala** [comment]
hi there, 
yes this thread is related to different point we discussed previously, so I try to follow ..
Until recently (https://github.com/fepegar/torchio/issues/116#issue-594892916) I  was not using Subject, so I am not sure, why to we need such a level.

I just resume to be sure I correctly understand
So your last proposition is changing the definition, but not the end result : dataset is still producing sample dictionary where the keys are the different image type. rigth ?

given that, it is more clear to keep the image name (which will be the key to get the image from the sample to the image level)

Now about adding extra parameters at the subject level, I wonder how you will use them, within a dataloader. everything is rearrange with batch size, so you can not match subject extra parameter ... no?
I thing you can not avoid the collate function, done by pytorch when creating the batch, since this is mandatory to have you extra parameter, still in correspondence with your batch size

does it make sense ?

**Comment by romainVala** [comment]
Although I do understand it make sense to have extra parameter at the subject level, since they may applied to all images, (for instance patient pathology, it not related to T1 and T2, but to both).

In your proposition the subject extra info is added to the sample, so it is fine (ie it will be collate in the batch). that's good

I am just not sure, why to remove the 'T1' key from the Image ...

**Comment by fepegar** [comment]
> I was not using Subject, so I am not sure why we need such a level

Mostly for classification/regression tasks of single numbers, I suppose. 

> So your last proposition is changing the definition, but not the end result : dataset is still producing sample dictionary where the keys are the different image type. right ?

Results should be the same, yes. The keys would be the image *name*.

> I am just not sure, why to remove the 'T1' key from the Image ...

If it's in the Subject, it doesn't need to be in the Image. I think this makes also more sense conceptually, as the image itself doesn't need to know its own name. Also it's consistent with how the sample dictionary is created. Subject and Image are just interfaces to create the dictionaries, so it seems logical that they somehow share part of the key-value structure.

File: torchio/data/images.py (hunk 2 above)
Comment (nwschurink) [reply torchio/data/images.py:44]: Having both options of `subject.kwargs['diagnose']` and `subject.diagnose` would be a solution. It would also be really easy to add.

As I'm reading [online](https://stackoverflow.com/questions/2033150/subclassing-dict-should-dict-init-be-called), I am starting to doubt a little bit on whether it makes sense to inherrit from `dict`. Apparently if you'd do, then it would be good practice to make the `Subject.__init__` function similar to how `dict.__init__` works. I think that would mean that the current initialization should be altered such that `Subject.__init__` should accept `key:value` pairs instead of positional arguments. Which would mean that it break all code written with previous versions right?

Comment (fepegar) [reply torchio/data/images.py:44]: Yes. That's what I propose in my [comment below](https://github.com/fepegar/torchio/pull/115#issuecomment-609696750). Let's continue the discussion there, please.

Don't worry about breaking the code: "[Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable](https://semver.org/#spec-item-4)".

If it makes the library better, I think some disruption is acceptable in this experimental phase.

**Comment by fepegar** [comment]
(See also discussion here: https://github.com/fepegar/torchio/pull/115#discussion_r404002739)

**Comment by romainVala** [comment]
I better understand, (sorry for my previous confusion)
Yes it make sense then,

**Comment by romainVala** [comment]
> ```python
> sample = {}
> for key, value in subject.items():
>     if isinstance(value, Image):
>         sample[key] = get_image_dict_from_image(value)
>     else:
>         sample[key] = value
> ```

where will be this part of the code  ?

**Comment by fepegar** [comment]
> where will be this part of the code ?

In this method:
https://github.com/fepegar/torchio/blob/22459930364374dd8be9d11ab5a0894722c7bb0d/torchio/data/images.py#L210-L221

**Comment by nwschurink** [comment]
Cool, this would indeed fix it! Should I add it to the PR? or do you edit it yourself?

**Comment by fepegar** [comment]
I'll give it a try.

**Comment by fepegar** [comment]
@romainVala @nwschurink can you take a look?

**Comment by romainVala** [comment]
Hi
great job, quite a lot of change, indeed, but always similar, so I am ready to do the same on my script.
seems good to me

**Comment by nwschurink** [comment]
Good morning,

Looks very good, onlys one thing that i'm wondering about. In `Subject.__repr__` a reference is made to `Subject.name` https://github.com/fepegar/torchio/pull/115/files/a92f075848fd7216704db4672699dbe43a649b33..3ceb0fcd191af252aac5ca934fca0653a4dc6095#diff-87328b3066b813d9ca439505a520b7b3R108, however, if `Subject.name` is undefined this will throw an exception.

I'm not sure if you'd want to enforce a name being set for each subject. If so something in the line of this could be added to the beginning of init just before this line https://github.com/fepegar/torchio/pull/115/files/a92f075848fd7216704db4672699dbe43a649b33..3ceb0fcd191af252aac5ca934fca0653a4dc6095#diff-87328b3066b813d9ca439505a520b7b3R99:

```
try:
    self.name = kwargs.pop('name')
except KeyError:
    raise KeyError('a "name" keyword argument must be supplied')
super().__init__(**kwargs)
```

Or if you'd don't want name to be a required argument you'd probably want to define it as an empty string (or delete the reference in `__repr__`) before you call `super().__init__`:

```
self.name = ''
super().__init__(**kwargs)
```

That way if `name` is defined as a `kwarg` it will overwrite `self.name`, and will be defined as an empty string otherwise.

**Comment by nwschurink** [comment]
Ah, never mind, my comment doesn't make sense.. `__name__` refers to the name of the module not the subject name. So the line will work ;) Learned something new.

Anyway, I'd say the code is good to go then
